test/integration/pause-resume-test.js

```diff
@@ -0,0 +1,246 @@
+      'select i from cte1 option (maxrecursion 0)';
+    request = new Request(sql, onRequestCompletion);
+    request.on('row', processRow);
+    connection.execSql(request);
+  }
+
+  function onRequestCompletion(err) {
+    test.ifError(err);
+    test.equal(rowsReceived, totalRows, 'Invalid row count.');
+    test.done();
+  }
+
+  function processRow(columns) {
+    if (paused) {
+      fail('Row received in paused state.');
+    }
+    rowsReceived++;
+    if (columns[0].value != rowsReceived) {
+      fail('Invalid row counter value, value=' + columns[0].value + ', expected=' + rowsReceived + '.');
+      return;
+    }
+    if (rowsReceived == Math.round(totalRows / 4)) {
+      pause();
+    }
+  }
+
+  function pause() {
+    if (debugMode) {
+      dumpStreamStates(connection);
+      console.log('Start pause.');
+    }
+    paused = true;
+    request.pause();
+    setTimeout(resume, delayTime);
+  }
+
+  function resume() {
+    if (debugMode) {
+      console.log('End pause.');
+      dumpStreamStates(connection);
+    }
+    verifyStreamStatesAfterPause();
+    paused = false;
+    request.resume();
+  }
+
+  function verifyStreamStatesAfterPause() {
+    const packetSize = connection.messageIo.packetSize();
+    const socketRs = connection.socket._readableState;
+    if (semver.gte(process.version, '0.12.18')) {
+      test.ok(!socketRs.flowing,
+        'Socket is not paused.');
+    }
+    test.ok(socketRs.length >= Math.min(socketRs.highWaterMark - 512, 0x4000),
+      'Socket does not feel backpressure.');
+    const packetTransformWs = connection.messageIo.packetStream._writableState;
+    const packetTransformRs = connection.messageIo.packetStream._readableState;
+    test.ok(!packetTransformRs.flowing,
+      'Packet transform is not paused.');
+    test.ok(packetTransformWs.length <= packetTransformWs.highWaterMark && packetTransformRs.length <= packetTransformRs.highWaterMark,
+      'Packet transform has large amount of data buffered.');
+    const tokenTransformWs = connection.tokenStreamParser.parser._writableState;
+    const tokenTransformRs = connection.tokenStreamParser.parser._readableState;
+    test.ok(!tokenTransformRs.flowing,
+      'Token transform is not paused.');
+    test.ok(tokenTransformWs.length <= tokenTransformWs.highWaterMark,
+      'Token transform input buffer overflow.');
+    test.ok(tokenTransformRs.length < packetSize / 3,
+      'Token transform output buffer has large amount of data buffered.');
+  }
+
+  function fail(msg) {
+    if (failed) {
+      return; }
+    failed = true;
+    test.ok(false, msg);
+    connection.close();
+  }
+};
+
+// This test reads only a few rows and makes a short pause after each row.
+// The test verifies that:
+//  - Pause/resume works correctly when applied after the last packet of a TDS
+//    message has already been dispatched by MessageIO.ReadablePacketStream.
+//    (EOM / packet.isLast() has already been detected when pause() is called.
```

Comment:
No, the end-bracket it at the end of the paragraph. But I will remove the brackets to make it clearer.